panther_driver/src/driver_node.py

```diff
@@ -335,22 +345,86 @@ def driverNode():
             except:
                 rospy.logwarn("Error reading rear right controller sdo")
 
+
             # position_FL / RK.encoder_resolution - > full wheel rotations
-            wheel_FL_ang_pos = 2 * math.pi * position_FL / RK.encoder_resolution  # radians
-            wheel_FR_ang_pos = 2 * math.pi * position_FR / RK.encoder_resolution
-            wheel_RL_ang_pos = 2 * math.pi * position_RL / RK.encoder_resolution
-            wheel_RR_ang_pos = 2 * math.pi * position_RR / RK.encoder_resolution
-
-            wheel_FL_ang_vel = (wheel_FL_ang_pos -
-                                wheel_FL_ang_pos_last) * dt_  # rad/s
-            wheel_FR_ang_vel = (wheel_FR_ang_pos - wheel_FR_ang_pos_last) * dt_
-            wheel_RL_ang_vel = (wheel_RL_ang_pos - wheel_RL_ang_pos_last) * dt_
-            wheel_RR_ang_vel = (wheel_RR_ang_pos - wheel_RR_ang_pos_last) * dt_
-
-            wheel_FL_ang_pos_last = wheel_FL_ang_pos
-            wheel_FR_ang_pos_last = wheel_FR_ang_pos
-            wheel_RL_ang_pos_last = wheel_RL_ang_pos
-            wheel_RR_ang_pos_last = wheel_RR_ang_pos
+            wheel_FL_ang_pos = 2.0 * math.pi * position_FL / RK.encoder_resolution / RK.gear_ratio  # radians
+            wheel_FR_ang_pos = 2.0 * math.pi * position_FR / RK.encoder_resolution / RK.gear_ratio
+            wheel_RL_ang_pos = 2.0 * math.pi * position_RL / RK.encoder_resolution / RK.gear_ratio
+            wheel_RR_ang_pos = 2.0 * math.pi * position_RR / RK.encoder_resolution / RK.gear_ratio
+
+
+            try:
+                velocity_FL = front_controller.sdo['Qry_ABSPEED'][2].raw
+            except:
+                rospy.logwarn("Error reading front left controller sdo")
+            try:
+                velocity_FR = front_controller.sdo['Qry_ABSPEED'][1].raw
+            except:
+                rospy.logwarn("Error reading front right controller sdo")
+            try:
+                velocity_RL = rear_controller.sdo['Qry_ABSPEED'][2].raw
+            except:
+                rospy.logwarn("Error reading rear left controller sdo")
+            try:  
+                velocity_RR = rear_controller.sdo['Qry_ABSPEED'][1].raw
+            except:
+                rospy.logwarn("Error reading rear right controller sdo")
```

Comment:
Same here and further (calculation of wheels velocity and torque). driverNode seems to be a huge blob, wee definitely should try to divide it and separate single responsibility methods.

Reply:
The refactor is must have.